README.md
[![Release](https://jitpack.io/v/MarcinOrlowski/datetimetemplate.svg)](https://jitpack.io/#MarcinOrlowski/DateTimeTemplate)
![Downloads](https://jitpack.io/v/MarcinOrlowski/DateTimeTemplate/month.svg)
[![Dependency Status](https://dependencyci.com/github/MarcinOrlowski/DateTimeTemplate/badge)](https://dependencyci.com/github/MarcinOrlowski/DateTimeTemplate)

DateTimeTemplate
================
`DateTimeTemplate` is flexible date/time formatting library with placeholders support.

![Demo app](img/demo.png)

Download demo application APK from [releases](https://github.com/MarcinOrlowski/DateTimeTemplate/releases)
section. Source code in project's `app/` module.

## Features ##

* Easy to use,
* Stable and production ready,
* Localization support,
* Lightweight,
* No extra dependencies,
* Free.

Installation
============

Edit your master `gradle.build` file and **add** `maven { url 'https://jitpack.io' }` to your current
`repositories` block content (if you use other jitpack hosted libraries, then this step can be skipped):

    allprojects {
      repositories {
        maven { url 'https://jitpack.io' }
        }
    }

Next, edit your **module**'s `build.gradle` and the following dependency:

    compile 'com.github.MarcinOrlowski:datetimetemplate:<VERSION>'

For recent value of `<VERSION>` consult [library releases](https://github.com/MarcinOrlowski/DateTimeTemplate/releases)
or jitpack
badge: [![Release](https://jitpack.io/v/MarcinOrlowski/datetimetemplate.svg)](https://jitpack.io/#MarcinOrlowski/datetimetemplate)


Placeholders
============

Your formatting string can contain anything you like, however certain sequences are considered placeholders, and will be replaced
by corresponding values. Non-placeholders are returned unprocessed

| Placeholder | Description |
|-------------|-------------|
| %yy%        | long year (i.e. "2009") |
| %y%        | short year (i.e. "09") |
| %MMM%        | long month name (i.e. "January") |
| %MM%        | abbreviated month name (i.e. "Jan") |
| %M%        | first letter of month name (i.e. "J") |
| %mm%        | zero prefixed 2 digit month number ("02" for Feb, "12" for Dec) |
| %m%        | month number as is ("2" for Feb, "12" for Dec) |
| %DDD%        | full day name (i.e. ""Saturday", "Sunday", "Monday") |
| %DD%        | abbreviated day name ("Sat", "Sun", "Mon") |
| %D%        | one letter day name ("S", "S", "M") |
| %dd%        | zero prefixed 2 digit day number ("01", "27") |
| %d%        | day number as is ("1", "27") |
| %dy%        | day number of the year (i.e. "250") |
| %dw%        | day number in week (i.e. "1" for Monday **if** weeks start on Mondays!) |
| %wm%        | week number of current month ("3" for 3rd week) |
| %wy%        | week number of the year ("3" for 3rd week, 47 for 47th) |
| %hh%        | current hour, zero prefixed, 24hrs clock (i.e. "01", "16") |
| %h%        | current hour, 24hrs clock (i.e. "1", "16") |
| %kk%        | current hour, zero prefixed, 12hrs clock (i.e. "01", "11") |
| %k%        | current hour, 12hrs clock (i.e. "1", "11") |
| %ii%        | current minute, zero prefixed (i.e. "01", "35") |
| %i%        | current minute, zero prefixed (i.e. "1", "35") |
| %AA%        | upper-cased AM/PM marker (i.e. "AM") |
| %A%        | upper-cased abbreviated AM/PM marker. "A" for "AM", "P" for "PM" |
| %aa%        | lower-cased am/pm marker (i.e. "am") |
| %a%        | lower-cased abbreviated AM/PM marker. "a" for "am", "p" for "pm" |
| %Aa%        | AM/PM marker with first letter uppercased (i.e. "Am"/"Pm") |

Examples
========

Format current date, time as it is in `GMT` zone, using 24hrs clock format:

    TimeZone tz = TimeZone.getTimeZone("GMT");
    Calendar c = new GregorianCalendar(tz);
    c.setTime(new Date());

    String formatted = DateTimeTemplate.format(c, "GMT: %DD% %hh%:%ii%");

would produce  `GMT: Mon 16:25`.

Format current time as it is in `Pacific Daylight Time` zone (`PDT` is 7 hours behind `GMT`), using 12hrs clock format with
abbreviated AM/PM marker:

    TimeZone tz = TimeZone.getTimeZone("GMT-0700");
    Calendar c = new GregorianCalendar(tz);
    c.setTime(new Date());

    String formatted = DateTimeTemplate.format(c, "Time: %k%:%ii%%a%");

would produce `Time: 3:25a`.

Format current time using device's time zone:

    TimeZone tz = TimeZone.getDefault();
    Calendar c = new GregorianCalendar(tz);
    c.setTime(new Date());

    String formatted = DateTimeTemplate.format(c, ...);

Formatting using locale of your choice:

    String formatted = DateTimeTemplate.format(cal, "Time: %k%:%ii%%a%", Locale.SIMPLIFIED_CHINESE);

Contributing
============

Please report any issue spotted using [GitHub's project tracker](https://github.com/MarcinOrlowski/DateTimeTemplate/issues).

If you'd like to contribute to the this project,
please [open new ticket](https://github.com/MarcinOrlowski/DateTimeTemplate/issues)
**before doing any work**. This will help us save your time in case I'd not be able to accept such changes. But if all is good and
clear then follow common routine:

* fork the project
* create new branch
* do your changes
* send pull request

License
=======

* Written and copyrighted &copy;2013-2022 by [Marcin Orlowski](https://marcinorlowski.com)
* DateTimeTemplate is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT)


CHANGES.md
# DateTimeTemplate for Android #

## CHANGELOG ##

1.5.0 (2022-11-21)
==================
* Demo app rewritten in Kotlin.
* Dropped ButterKnife in favor of data binding.
* Updated dependencies to recent versions.

1.4.0 (2020-06-04)
==================
* Updated demo app to use AndroidX
* Updated dependencies to recent versions

1.3.1 (2018-03-11)
==================
* Fixed bug in `format()` not replacing placeholders

1.3.0 (2018-03-11) (broken!)
==================
* Added support for `%Aa%` placeholder

1.2.4 (2018-02-23)
==================
* Added date/time pickers to demo app
* [#10] `%k%` and `%kk%` produced 0 instead of 12 for noon and midnight

v1.2.3 (2017-07-05)
===================
* Updated dependencies

v1.2.2 (2017-05-26)
===================
* Disabled proguard for the library

v1.2.1 (2017-05-25)
===================
* [#6] Fixed proguard config file

v1.2.0 (2017-05-24)
===================
* [#2] Fixed `%mm%` producing wrong results
* [#4] Calendar's TZ was not used for localized placeholders (i.e. day names)

v1.1.0 (2017-03-09)
===================
* You can now pass `Locale` of your choice to `format()` when needed

v1.0.1 (2017-03-08)
===================
* Lowered `minSdk` to `ICE_CREAM_SANDWICH` (`14`)

v1.0.0 (2017-03-08)
===================
* First public release


lib/src/test/java/marcinorlowski/com/datetimetemplate/ExampleUnitTest.java
package marcinorlowski.com.datetimetemplate;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
	@Test
	public void addition_isCorrect() throws Exception {
		assertEquals(4, 2 + 2);
	}
}


app/src/main/java/com/marcinorlowski/datetimetemplate/demo/MainActivity.kt
package com.marcinorlowski.datetimetemplate.demo

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.app.Dialog
import android.app.TimePickerDialog
import android.app.TimePickerDialog.OnTimeSetListener
import android.os.Bundle
import android.text.format.DateFormat
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.DatePicker
import android.widget.TimePicker
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.DialogFragment
import com.marcinorlowski.datetimetemplate.DateTimeTemplate
import com.marcinorlowski.datetimetemplate.demo.databinding.ActivityMainBinding
import java.util.*

/** ******************************************************************************
 *
 * DateTimeTemplate
 * Flexible date/time formatting library with placeholders support.
 *
 * @author Marcin Orlowski <mail (#) marcinOrlowski (.) com>
 * @copyright 2013-2022 Marcin Orlowski
 * @license http://www.opensource.org/licenses/mit-license.php MIT
 * @link https://github.com/MarcinOrlowski/DateTimeTemplate
 *
 * *************************************************************************** **/

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        mCalendar.time = Date()
        mHour = mCalendar[Calendar.HOUR_OF_DAY]
        mMinute = mCalendar[Calendar.MINUTE]
        mYear = mCalendar[Calendar.YEAR]
        mMonth = mCalendar[Calendar.MONTH]
        mDay = mCalendar[Calendar.DAY_OF_MONTH]

        binding.update.setOnClickListener { update() }
        binding.time.setOnClickListener { pickTime() }
        binding.date.setOnClickListener { pickDate() }
    }

    private fun update() {
        val windowToken = window.decorView.rootView.windowToken
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(windowToken, 0)
        binding.root.clearFocus()
        binding.text.text = DateTimeTemplate.format(mCalendar, binding.format.text.toString())
        binding.resultCard.visibility = View.VISIBLE
    }

    // --------------------------------------------------------------------------------------------------------------------------

    private fun pickTime() {
        TimePickerFragment().show(supportFragmentManager, "timePicker")
    }

    class TimePickerFragment : DialogFragment(), OnTimeSetListener {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            return TimePickerDialog(activity, this, mHour, mMinute, DateFormat.is24HourFormat(activity))
        }

        override fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int) {
            mHour = hourOfDay
            mMinute = minute
            updateCalendar()
        }
    }

    // --------------------------------------------------------------------------------------------------------------------------

    fun pickDate() {
        DatePickerFragment().show(supportFragmentManager, "datePicker")
    }

    class DatePickerFragment : DialogFragment(), OnDateSetListener {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            return DatePickerDialog(requireActivity(), this, mYear, mMonth, mDay)
        }

        override fun onDateSet(view: DatePicker, year: Int, month: Int, day: Int) {
            mYear = year
            mMonth = month
            mDay = day
            updateCalendar()
        }
    }

    companion object {
        val mCalendar: Calendar = GregorianCalendar(TimeZone.getDefault())
        protected fun updateCalendar() {
            mCalendar[mYear, mMonth, mDay, mHour] = mMinute
        }

        protected var mHour = 0
        protected var mMinute = 0
        protected var mYear = 0
        protected var mMonth = 0
        protected var mDay = 0
    }
}


app/src/main/java/com/marcinorlowski/datetimetemplate/demo/MainActivity.java
package com.marcinorlowski.datetimetemplate.demo;

import android.app.DatePickerDialog;
import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.os.Bundle;
import android.os.IBinder;
import android.text.format.DateFormat;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.TimePicker;

import com.marcinorlowski.datetimetemplate.DateTimeTemplate;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.TimeZone;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.DialogFragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * DateTimeTemplate - a flexible date/time formatting library with placeholders support.
 *
 * @author Marcin Orlowski <mail (#) marcinOrlowski (.) com>
 * @copyright 2013-2020 Marcin Orlowski
 * @license http://www.opensource.org/licenses/mit-license.php MIT
 * @link https://github.com/MarcinOrlowski/DateTimeTemplate
 */

public class MainActivity extends AppCompatActivity
{
	@BindView (R.id.root)
	View mRootContainer;

	@BindView (R.id.format)
	EditText mFormat;

	@BindView (R.id.text)
	TextView mText;

	@BindView (R.id.resultCard)
	View mResultCard;

	@BindView (R.id.time)
	Button mTime;

	@BindView (R.id.date)
	Button mDate;

	protected Unbinder mButterKnifeUnbinder;

	final static Calendar mCalendar = new GregorianCalendar(TimeZone.getDefault());

	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		mCalendar.setTime(new Date());

		mHour = mCalendar.get(Calendar.HOUR_OF_DAY);
		mMinute = mCalendar.get(Calendar.MINUTE);
		mYear = mCalendar.get(Calendar.YEAR);
		mMonth = mCalendar.get(Calendar.MONTH);
		mDay = mCalendar.get(Calendar.DAY_OF_MONTH);

		mButterKnifeUnbinder = ButterKnife.bind(this);
	}

	@Override
	protected void onDestroy()
	{
		mButterKnifeUnbinder.unbind();
		super.onDestroy();
	}

	@OnClick (R.id.update)
	public void update()
	{
		IBinder windowToken = getWindow().getDecorView().getRootView().getWindowToken();
		InputMethodManager imm = (InputMethodManager)getSystemService(Context.INPUT_METHOD_SERVICE);
		imm.hideSoftInputFromWindow(windowToken, 0);
		mRootContainer.clearFocus();

		mText.setText(DateTimeTemplate.format(mCalendar, mFormat.getText().toString()));

		mResultCard.setVisibility(View.VISIBLE);
	}

	protected static void updateCalendar()
	{
		mCalendar.set(mYear, mMonth, mDay, mHour, mMinute);
	}

	// --------------------------------------------------------------------------------------------------------------------------

	@OnClick (R.id.time)
	public void pickTime()
	{
		DialogFragment newFragment = new TimePickerFragment();
		newFragment.show(getSupportFragmentManager(), "timePicker");
	}

	protected static int mHour, mMinute;

	public static class TimePickerFragment extends DialogFragment implements TimePickerDialog.OnTimeSetListener
	{
		@Override
		@NonNull
		public Dialog onCreateDialog(Bundle savedInstanceState)
		{
			return new TimePickerDialog(getActivity(), this, mHour, mMinute, DateFormat.is24HourFormat(getActivity()));
		}

		public void onTimeSet(TimePicker view, int hourOfDay, int minute)
		{
			mHour = hourOfDay;
			mMinute = minute;

			updateCalendar();
		}
	}

	// --------------------------------------------------------------------------------------------------------------------------

	@OnClick (R.id.date)
	public void pickDate()
	{
		DialogFragment newFragment = new DatePickerFragment();
		newFragment.show(getSupportFragmentManager(), "datePicker");
	}

	protected static int mYear, mMonth, mDay;

	public static class DatePickerFragment extends DialogFragment implements DatePickerDialog.OnDateSetListener
	{
		@Override
		@NonNull
		public Dialog onCreateDialog(Bundle savedInstanceState)
		{
			return new DatePickerDialog(getActivity(), this, mYear, mMonth, mDay);
		}

		public void onDateSet(DatePicker view, int year, int month, int day)
		{
			mYear = year;
			mMonth = month;
			mDay = day;

			updateCalendar();
		}
	}

}


lib/src/main/java/com/marcinorlowski/datetimetemplate/DateTimeTemplate.java
package com.marcinorlowski.datetimetemplate;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.EnumMap;
import java.util.Locale;

/** ******************************************************************************
 *
 * DateTimeTemplate
 * Flexible date/time formatting library with placeholders support.
 *
 * @author Marcin Orlowski <mail (#) marcinOrlowski (.) com>
 * @copyright 2013-2022 Marcin Orlowski
 * @license http://www.opensource.org/licenses/mit-license.php MIT
 * @link https://github.com/MarcinOrlowski/DateTimeTemplate
 *
 * *************************************************************************** **/

public class DateTimeTemplate
{
	// supported placeholders
	enum Placeholder
	{
		yy, y,
		MMM, MM, M, mm, m,
		DDD, DD, D,
		dd, d, dy, dw,
		wm, wy,

		hh, h, kk, k,
		ii, i,

		AA, A, aa, a, Aa
	}

	/**
	 * Processed given formatting string, substituting all supported placeholders.
	 *
	 * @param cal    Calendar object to use as date/time source
	 * @param format Formatting string
	 *
	 * @return processed date/time string
	 */
	public static String format(Calendar cal, String format)
	{
		return format(cal, format, false);
	}

	/**
	 * Processed given formatting string, substituting all supported placeholders.
	 *
	 * @param cal          Calendar object to use as date/time source
	 * @param format       Formatting string
	 * @param forceEnglish enforces use of English language instead of default locale (whenever applicable)
	 *
	 * @return processed date/time string
	 */
	@SuppressWarnings ("WeakerAccess")
	public static String format(Calendar cal, String format, @SuppressWarnings ("SameParameterValue") boolean forceEnglish)
	{
		Locale locale = Locale.ENGLISH;

		// Checks if formatter supports current system locale or we need to fallback to English
		// forces english in string Placeholder like %MMM%. Also works around HTC Desire bug on
		// Froyo or in general workaround for locale not supported by formatter.
		if (!forceEnglish) {
			String currentLanguage = Locale.getDefault().getLanguage();

			Locale[] availableLocales = SimpleDateFormat.getAvailableLocales();
			for (Locale availableLocale : availableLocales) {
				if (availableLocale.getLanguage().equals(currentLanguage)) {
					locale = availableLocale;
					break;
				}
			}
		}

		return format(cal, format, locale);
	}

	/**
	 * Processed given formatting string, substituting all supported placeholders.
	 *
	 * @param cal    Calendar object to use as date/time source
	 * @param format Formatting string
	 * @param locale Locale to use for month, day names
	 *
	 * @return processed date/time string
	 */
	@SuppressWarnings ("WeakerAccess")
	public static String format(Calendar cal, String format, Locale locale)
	{
		EnumMap<Placeholder, String> map = new EnumMap<>(Placeholder.class);

		Date date = new java.util.Date(cal.getTimeInMillis());
		SimpleDateFormat formatter = new SimpleDateFormat("", locale);
		formatter.setTimeZone(cal.getTimeZone());

		// this valid for 2012 only, where 1st is on Sunday
		// some tweaks to make WY work correctly and show "1" instead of "52" as used in 2012
		// http://stackoverflow.com/questions/7299621/android-calendar-problem-with-day-of-the-week
		Calendar calendar = formatter.getCalendar();
		calendar.setMinimalDaysInFirstWeek(1);
		calendar.setFirstDayOfWeek(Calendar.SUNDAY);
		calendar.set(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH));


		// %Y% - long year (2010)
		// %y% - short year (10)
		map.put(Placeholder.yy, Integer.toString(cal.get(Calendar.YEAR)));
		map.put(Placeholder.y, map.get(Placeholder.yy).substring(2));

		// %MMM% - long month name (January)
		// %MM%	 - abbreviated month name (Jan)
		// %M%	 - first letter of month name (J)
		formatter.applyLocalizedPattern("MMMM");
		map.put(Placeholder.MMM, formatter.format(date));

		formatter.applyLocalizedPattern("MMM");
		map.put(Placeholder.MM, formatter.format(date));
		map.put(Placeholder.M, map.get(Placeholder.MM).substring(0, 1));

		// %mm%	- zero prefixed 2 digit month number (02 for Feb, 12 for Dec)
		// %m%	- month number as is (2 for Feb, 12 for Dec)
		formatter.applyLocalizedPattern("MM");
		map.put(Placeholder.mm, formatter.format(date));

		formatter.applyLocalizedPattern("M");
		map.put(Placeholder.m, formatter.format(date));


		// %DDD% - full day name (Saturday, Monday etc)
		// %DD%  - abbreviated day name (Sat, Mon)
		// %D%   - one letter, abbreviated day name (S, M)
		formatter.applyLocalizedPattern("EEEE");
		map.put(Placeholder.DDD, formatter.format(date));

		formatter.applyLocalizedPattern("EEE");
		String abrvDayName = formatter.format(date);
		map.put(Placeholder.DD, abrvDayName);
		map.put(Placeholder.D, abrvDayName.substring(0, 1).toUpperCase());

		// %dd% - zero prefixed 2 digit day number (01 for 1st, 27 for 27th)
		// %d%  - day number as is (1 for 1st, 27 for 27th)
		// %dy% - day number of the year (i.e. 250)
		// %dw% - day number in week (i.e. 1 for Monday if weeks start on Mondays)
		formatter.applyLocalizedPattern("dd");
		map.put(Placeholder.dd, formatter.format(date));

		formatter.applyLocalizedPattern("d");
		map.put(Placeholder.d, formatter.format(date));

		formatter.applyLocalizedPattern("D");
		map.put(Placeholder.dy, formatter.format(date));

		map.put(Placeholder.dw, Integer.toString(cal.get(Calendar.DAY_OF_WEEK)));


		// %wm% - week number of current month (3 for 3rd week)
		// %wy% - week number of the year (3 for 3rd week, 47 for 47th)
		formatter.applyLocalizedPattern("W");
		map.put(Placeholder.wm, formatter.format(date));

		formatter.applyLocalizedPattern("w");
		map.put(Placeholder.wy, formatter.format(date));


		// %hh% - current hour, zero prefixed, 24hrs clock (i.e. "01", "16")
		// %h% - current hour, 24hrs clock (i.e. "1", "16")
		int hour = cal.get(Calendar.HOUR_OF_DAY);
		map.put(Placeholder.hh, (hour < 10) ? "0" + hour : Integer.toString(hour));
		map.put(Placeholder.h, Integer.toString(hour));

		// %kk% - current hour, zero prefixed, 12hrs clock (i.e. "01", "11")
		// %k%  - current hour, 12hrs clock (i.e. "1", "11")
		hour = cal.get(Calendar.HOUR);
		if (hour == 0) {
			hour = 12;
		}
		map.put(Placeholder.kk, (hour < 10) ? "0" + hour : Integer.toString(hour));
		map.put(Placeholder.k, Integer.toString(hour));

		// %ii% - current minute, zero prefixed (i.e. "01", "35")
		// %i%  - current minute, zero prefixed (i.e. "1", "35")
		int minute = cal.get(Calendar.MINUTE);
		map.put(Placeholder.ii, (minute < 10) ? "0" + minute : Integer.toString(minute));
		map.put(Placeholder.i, Integer.toString(minute));

		// %AA% - upper-cased AM/PM marker (i.e. "AM")
		// %A%  - upper-cased abbreviated AM/PM marker. "A" for "AM", "P" for "PM"
		// %aa% - lower-cased am/pm marker (i.e. "am")
		// %a%  - lower-cased abbreviated AM/PM marker. "a" for "am", "p" for "pm"
		// %Aa% - AM/PM marker with first letter uppercased (i.e. "Am"/"Pm")
		int ampm = cal.get(Calendar.AM_PM);
		map.put(Placeholder.AA, (ampm == Calendar.AM) ? "AM" : "PM");
		map.put(Placeholder.A, (ampm == Calendar.AM) ? "A" : "P");
		map.put(Placeholder.aa, map.get(Placeholder.AA).toLowerCase());
		map.put(Placeholder.a, map.get(Placeholder.A).toLowerCase());
		map.put(Placeholder.Aa, (ampm == Calendar.AM) ? "Am" : "Pm");


		// placeholder substitution...
		String pattern, result = format;
		for (Placeholder placeholderKey : Placeholder.values()) {
			pattern = "%" + placeholderKey.name() + "%";
			result = result.replaceAll(pattern, map.get(placeholderKey));
		}

		return result;
	}
}
